src/components/Editor/Runners/PythonRunner/PythonRunner.jsx

```diff
@@ -1,462 +1,58 @@
-                <div className="react-tabs__tab-container">
-                  <TabList>
-                    <Tab key={0}>
-                      <span className="react-tabs__tab-text">
-                        {t("output.visualOutput")}
-                      </span>
-                    </Tab>
-                  </TabList>
-                  {!isEmbedded && hasVisualOutput ? <OutputViewToggle /> : null}
-                  {!isEmbedded && isMobile ? <RunnerControls skinny /> : null}
-                </div>
-                <TabPanel key={0}>
-                  <VisualOutputPane />
-                </TabPanel>
-              </Tabs>
-            </div>
-          ) : null}
-          <div className="output-panel output-panel--text">
-            <Tabs forceRenderTabPanel={true}>
-              <div className="react-tabs__tab-container">
-                <TabList>
-                  <Tab key={0}>
-                    <span className="react-tabs__tab-text">
-                      {t("output.textOutput")}
-                    </span>
-                  </Tab>
-                </TabList>
-                {!hasVisualOutput && !isEmbedded && isMobile ? (
-                  <RunnerControls skinny />
-                ) : null}
-              </div>
-              <ErrorMessage />
-              <TabPanel key={0}>
-                <pre
-                  className={`pythonrunner-console pythonrunner-console--${settings.fontSize}`}
-                  onClick={shiftFocusToInput}
-                  ref={output}
-                ></pre>
-              </TabPanel>
-            </Tabs>
-          </div>
-        </>
-      ) : (
-        <Tabs forceRenderTabPanel={true} defaultIndex={hasVisualOutput ? 0 : 1}>
-          <div className="react-tabs__tab-container">
-            <TabList>
-              {hasVisualOutput ? (
-                <Tab key={0}>
-                  <span className="react-tabs__tab-text">
-                    {t("output.visualOutput")}
-                  </span>
-                </Tab>
-              ) : null}
-              <Tab key={1}>
-                <span className="react-tabs__tab-text">
-                  {t("output.textOutput")}
-                </span>
-              </Tab>
-            </TabList>
-            {!isEmbedded && hasVisualOutput ? <OutputViewToggle /> : null}
-            {!isEmbedded && isMobile ? <RunnerControls skinny /> : null}
-          </div>
-          <ErrorMessage />
-          {hasVisualOutput ? (
-            <TabPanel key={0}>
-              <VisualOutputPane />
-            </TabPanel>
-          ) : null}
-          <TabPanel key={1}>
-            <pre
-              className={`pythonrunner-console pythonrunner-console--${settings.fontSize}`}
-              onClick={shiftFocusToInput}
-              ref={output}
-            ></pre>
-          </TabPanel>
-        </Tabs>
-      )}
-    </div>
+    <>
+      <PyodideRunner active={usePyodide} />
+      <SkulptRunner active={!usePyodide} />
```

Comment:
Ah yes, strangely enough, these lines are the fix to the p5 canvas problem 😅 Because we're switching to skulpt for p5, there's no problem with there being multiple canvases if we keep the PyodideRunner there... But if we fully switch runners like with the terniary operator it will create multiple p5 canvases if the user does too much switching 😬 But just hiding the irrelevant runner with CSS means that every time the user switches back to skulpt the old p5 canvas is still there so it doesn't create another one.

Reply:
That makes complete sense 😄 so just ignore me and thanks for the details